compiling on ppc64le - compile error - char is unsigned

gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-4)

When compiling on Power8 (Centos7) I got the following compile error:

<pre>
[ 19%] Building C object storage/myisam/CMakeFiles/myisam_embedded.dir/ft_stopwords.c.o                                                                           
In file included from /home/danielgb/mysql-5.6/sql/sql_lex.h:25:0,
                 from /home/danielgb/mysql-5.6/sql/sql_class.h:546,
                 from /home/danielgb/mysql-5.6/storage/archive/ha_archive.cc:21:
/home/danielgb/mysql-5.6/sql/item.h:653:21: error: enumerator value -1 is too large for underlying type ‘char’
     FUNC_UNKNOWN = -1,
                     ^
/home/danielgb/mysql-5.6/sql/item.h:654:5: error: enumerator value 256 is too large for underlying type ‘char’
     FUNC_SET,
     ^
In file included from /home/danielgb/mysql-5.6/sql/sql_lex.h:25:0,
                 from /home/danielgb/mysql-5.6/sql/sql_class.h:546,
                 from /home/danielgb/mysql-5.6/sql/sql_base.h:21,
                 from /home/danielgb/mysql-5.6/sql/rpl_record.h:29,
                 from /home/danielgb/mysql-5.6/sql/log_event.h:44,
                 from /home/danielgb/mysql-5.6/sql/binlog.h:20,
                 from /home/danielgb/mysql-5.6/sql/rpl_slave.h:49,
                 from /home/danielgb/mysql-5.6/sql/sql_servers.h:20,
                 from /home/danielgb/mysql-5.6/storage/federated/ha_federated.cc:376:
/home/danielgb/mysql-5.6/sql/item.h:653:21: error: enumerator value -1 is too large for underlying type ‘char’
     FUNC_UNKNOWN = -1,
                     ^
/home/danielgb/mysql-5.6/sql/item.h:654:5: error: enumerator value 256 is too large for underlying type ‘char’
     FUNC_SET,
     ^
</pre>


char is unsigned by default on Power (and signed on other arches). Being explicit in this case to signed fixes the compile error:

Fix:

<pre>
diff --git a/sql/item.h b/sql/item.h
index b04d4c4..b08c5bb 100644
--- a/sql/item.h
+++ b/sql/item.h
@@ -649,7 +649,7 @@ typedef void (*Cond_traverser) (const Item *item, void *arg);
 // This class is the arguments for the partial update
 class Save_in_field_args : public Sql_alloc {
  public:
-  enum class FuncType : char {
+  enum class FuncType : signed char {
     FUNC_UNKNOWN = -1,
     FUNC_SET,
     FUNC_UNSET,
</pre>


Relates to commit 4d48e92a8e71411b5a1f9c51fcfd758deffbdf71

--- a/sql/item.h
+++ b/sql/item.h
@@ -649,7 +649,7 @@ typedef void (*Cond_traverser) (const Item *item, void *arg);
 // This class is the arguments for the partial update
 class Save_in_field_args : public Sql_alloc {
  public:
-  enum class FuncType : char {
+  enum class FuncType : signed char {
     FUNC_UNKNOWN = -1,
     FUNC_SET,
     FUNC_UNSET,

